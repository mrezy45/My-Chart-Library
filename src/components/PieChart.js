import React from 'react';
import '../styles/ChartStyles.css'

const PieChart = ({ data, colors = [] }) => {
  const total = data.reduce((sum, d) => sum + d.value, 0);
  let cumulativeValue = 0;

  return (
    <div className="chart-container">
      <svg width="200px" height="200px" viewBox="-100 -100 200 200">
        {data.map((d, i) => {
          const startAngle = (cumulativeValue / total) * 2 * Math.PI;
          cumulativeValue += d.value;
          const endAngle = (cumulativeValue / total) * 2 * Math.PI;

          const largeArcFlag = d.value / total > 0.5 ? 1 : 0;
          const [startX, startY] = [Math.cos(startAngle) * 100, Math.sin(startAngle) * 100];
          const [endX, endY] = [Math.cos(endAngle) * 100, Math.sin(endAngle) * 100];

          return (
            <path
              key={i}
              d={`M 0 0 L ${startX} ${startY} A 100 100 0 ${largeArcFlag} 1 ${endX} ${endY} Z`}
              fill={colors[i % colors.length] || 'steelblue'}
            />
          );
        })}
      </svg>
    </div>
  );
};

export default PieChart;