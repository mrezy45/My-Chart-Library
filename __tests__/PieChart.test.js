import React from 'react';
import { render } from '@testing-library/react';
import PieChart from '../src/components/PieChart';

describe('PieChart Component', () => {
  it('renders without crashing', () => {
    const data = [{ value: 50 }, { value: 80 }];
    const { container } = render(<PieChart data={data} colors={['#ff0000', '#00ff00']} />);
    expect(container).toMatchSnapshot();
  });

  it('renders the correct number of paths', () => {
    const data = [{ value: 50 }, { value: 80 }, { value: 30 }];
    const { container } = render(<PieChart data={data} colors={['#ff0000', '#00ff00']} />);
    const paths = container.querySelectorAll('path');
    expect(paths.length).toBe(data.length);
  });
});
