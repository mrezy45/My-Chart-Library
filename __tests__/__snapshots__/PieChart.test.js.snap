// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart Component renders without crashing 1`] = `
<div>
  <div
    class="chart-container"
  >
    <svg
      height="200px"
      viewBox="-100 -100 200 200"
      width="200px"
    >
      <path
        d="M 0 0 L 100 0 A 100 100 0 0 1 -74.85107481711012 66.31226582407952 Z"
        fill="#ff0000"
      />
      <path
        d="M 0 0 L -74.85107481711012 66.31226582407952 A 100 100 0 1 1 100 -2.4492935982947064e-14 Z"
        fill="#00ff00"
      />
    </svg>
  </div>
</div>
`;
